camel-core/src/main/java/org/apache/camel/processor/Throttler.java

```diff
@@ -235,63 +283,122 @@ protected void enqueuePermit(final ThrottlePermit permit, final Exchange exchang
     /**
      * Evaluates the maxRequestsPerPeriodExpression and adjusts the throttle rate up or down.
      */
-    protected void calculateAndSetMaxRequestsPerPeriod(final Exchange exchange) throws Exception {
+    protected void calculateAndSetMaxRequestsPerPeriod(DelayQueue<ThrottlePermit> delayQueue, final Exchange exchange, final Integer key) throws Exception {
         Integer newThrottle = maxRequestsPerPeriodExpression.evaluate(exchange, Integer.class);
 
         if (newThrottle != null && newThrottle < 0) {
             throw new IllegalStateException("The maximumRequestsPerPeriod must be a positive number, was: " + newThrottle);
         }
 
-        synchronized (this) {
-            if (newThrottle == null && throttleRate == 0) {
+        Object lockOnSync = null;
+        Integer currentThrottleRate = throttleRate;
+        if (correlationExpression != null) {
+            currentThrottleRate = throttleRatesMap.get(key);
+            lockOnSync = new Integer(currentThrottleRate.intValue());
```

Comment:
Can we use the throttleRate as the lock here? I don't think create a new Integer can do the job.

Reply:
Indeed, creating a new Integer each time can't do the job.
Let's use throttleRate as the lock.
My understanding is that all groups will share the same lock, but optimization could come later on.

Reply:
@WillemJiang and @aldettinger , guys can you please be more explicit about this subject? I am afraid i did not get what you mean.

Reply:
Our suggestion is using the currentThrottleRate as the lock, and if you create a new Integer with the currentThrottleRate intValue, you always get a new Integer instance for the lockOnSync variable. It doesn't  block the thread which try to access same ThrottleRate at the same time.

Reply:
I got that but that was the point of @aldettinger's first suggestion to make a defensive copy due to not being sure of the reference returned by evaluate could be used as a lock by another thread for another purpose or not. I think that led me to think like that. I think the correct way of managing synchronised block will be like this
    1- If no grouping applied, throttler itself should be used as lock because all threads will be accessing volatile throttleRate and modify it.
    2- If grouping applied, not the value from throttleRates map should be used as lock but the key would be a better choice.

that was what i already mentioned but reflected in an unintended way to implementation.
https://github.com/apache/camel/pull/2376#discussion_r195896140

Reply:
@onderson
Sorry for confusion. I think it will be clearer expressing my preferences below:
1) Make a defensive copy of the key only once, and lock on that same object for the whole group life
2) Lock on the value, currentThrottleRate
3) Lock on throttleRate (group would have been mixed)

Reply:
That's all right. It was a good catch.
I disagree about defensive copy of the key because those keys (key of throttleRatesMap) are already private and non-accessible variables, and there is no other place those keys used by any other threading mechanism. 
option 2 and 3 and are not good choices already IMHO whereas we can use evaluated value of correlationExpression as the lock for synchronization.